{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "lines_to_next_cell": 0
   },
   "source": [
    "# plan\n",
    "\n",
    "- Get hit scores from the jsons.\n",
    "  - need to get the sequences (from alignment file or matrix files) from there as well to build the logo.\n",
    "- apply the position weighting to the scores.\n",
    "\n",
    "\n",
    "steps:\n",
    "- construct table with score lists\n",
    "\n",
    "\n",
    "- import table\n",
    "- add flanking sequence\n",
    "- add score lists (by score_key/level)\n",
    "- use weights to get ave scores from score lists\n",
    "- get difference between aln and pairwise\n",
    "- apply a few filters to get likely interesting hits\n",
    "- plot logos with 5 flanking residues\n",
    "  - aln w/ gaps | aln w/o gaps | pairwise\n",
    "-\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#\n",
    "import json\n",
    "import os\n",
    "import re\n",
    "import sys\n",
    "from dataclasses import dataclass\n",
    "from pathlib import Path\n",
    "\n",
    "import local_conservation_analysis_pipeline.group_conservation_objects as group_tools\n",
    "import local_conservation_scores.tools.pairwise_tools as pairwise_tools\n",
    "import local_conservation_scores.tools.score_plots as score_plots\n",
    "import local_seqtools.general_utils as tools\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import Align, AlignIO, Seq, SeqIO\n",
    "from local_seqtools import pssms\n",
    "\n",
    "\n",
    "plt.style.use(\"custom_standard\")\n",
    "# plt.style.use('custom_small')\n",
    "import seaborn as sns\n",
    "\n",
    "# pd.options.plotting.backend = \"plotly\"\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_jsonfile_to_relative(json_file):\n",
    "    return json_file.replace(\"/home/jch/Documents/07-pt1_pt2_combined/\", \"../../\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "position_weights = {\n",
    "    \"DOC_WW_Pin1_4\": np.array([0, 0, 0, 1, 1, 0]),  # ...([ST])P.\n",
    "    \"LIG_AP2alpha_2\": np.array([1, 1, 1]),  # DP[FW]\n",
    "    \"LIG_EH_1\": np.array([0, 1, 1, 1, 0]),  # .NPF.\n",
    "    \"LIG_SH2_GRB2like\": np.array([1, 1, 1, 0]),  # (Y)([EDST]|[MLIVAFYHQW])N.\n",
    "    \"LIG_SH3_CIN85_PxpxPR_1\": np.array([1, 0, 1, 0, 1, 1]),  # P.[AP].PR\n",
    "    \"enah_LPPPP_FPPPP\": np.array([2, 1, 0, 1, 1]),  # [FWYL]P.[AFILTVYWP]P\n",
    "    \"TRAF6\": np.array([0, 0, 0, 1, 0, 1, 0, 0, 1]),  # ...P.E..[FYWDE]\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.average([3, 4, 2, 2], weights=[0, 0, 1, 1])\n",
    "# np.average([3,4,2,2], weights=[1,1,1,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_file = (\n",
    "    \"../../benchmark/benchmark_v3/p3_conservation/benchmark_table_renamed_ANNOTATED.csv\"\n",
    ")\n",
    "df = pd.read_csv(table_file)\n",
    "df = df[\n",
    "    df[\"ELM_motif_class\"] != \"LIG_14-3-3_CanoR_1\"\n",
    "]  # this motif has a variable length regex and so it's more difficult to apply any position weighting\n",
    "df = df[\n",
    "    [\n",
    "        \"reference_index\",\n",
    "        \"ELM_motif_class\",\n",
    "        \"Organism\",\n",
    "        \"UniprotID\",\n",
    "        \"regex\",\n",
    "        \"hit_sequence\",\n",
    "        \"gene_id\",\n",
    "        \"hit start position\",\n",
    "        \"hit end position\",\n",
    "        \"verified interaction\",\n",
    "        \"name\",\n",
    "        \"json_file\",\n",
    "        \"critical_error\",\n",
    "    ]\n",
    "]\n",
    "df = df[df[\"critical_error\"].isna()]\n",
    "df[\"json_file\"] = df[\"json_file\"].apply(convert_jsonfile_to_relative)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add flanking sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import local_env_variables.env_variables as env\n",
    "\n",
    "data_all_seqrecords_dict = env.load_data_all_odb_seqs()\n",
    "df[\"odb_seq\"] = df[\"gene_id\"].apply(\n",
    "    lambda x: (\n",
    "        str(data_all_seqrecords_dict[x].seq)\n",
    "        if x in data_all_seqrecords_dict.keys()\n",
    "        else False\n",
    "    )\n",
    ")\n",
    "df[\"flanked_hit\"] = df.apply(\n",
    "    lambda x: tools.pad_with_aas_or_gaps(\n",
    "        x[\"odb_seq\"], x[\"hit start position\"], x[\"hit end position\"] + 1, flank=5\n",
    "    ),\n",
    "    axis=1,\n",
    "    result_type=\"expand\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Eukaryota aln_property_entropy\n",
      "Eukaryota aln_shannon_entropy\n",
      "Eukaryota fragpair_gapless_lf5_rf5_edssmat50\n",
      "Eukaryota fragpair_gapless_lf0_rf0_edssmat50\n",
      "Eukaryota frag_pairwise_gapless_embedding_lf5_rf5\n",
      "Eukaryota frag_pairwise_gapless_embedding_lf0_rf0\n",
      "Metazoa aln_property_entropy\n",
      "Metazoa aln_shannon_entropy\n",
      "Metazoa fragpair_gapless_lf5_rf5_edssmat50\n",
      "Metazoa fragpair_gapless_lf0_rf0_edssmat50\n",
      "Metazoa frag_pairwise_gapless_embedding_lf5_rf5\n",
      "Metazoa frag_pairwise_gapless_embedding_lf0_rf0\n",
      "Vertebrata aln_property_entropy\n",
      "Vertebrata aln_shannon_entropy\n",
      "Vertebrata fragpair_gapless_lf5_rf5_edssmat50\n",
      "Vertebrata fragpair_gapless_lf5_rf5_blosum62\n",
      "Vertebrata fragpair_gapless_lf2_rf2_edssmat50\n",
      "Vertebrata fragpair_gapless_lf10_rf10_edssmat50\n",
      "Vertebrata fragpair_gapless_lf0_rf0_edssmat50\n",
      "Vertebrata frag_pairwise_gapless_embedding_lf5_rf5\n",
      "Vertebrata frag_pairwise_gapless_embedding_lf0_rf0\n",
      "Tetrapoda aln_property_entropy\n",
      "Tetrapoda aln_shannon_entropy\n",
      "Tetrapoda fragpair_gapless_lf5_rf5_edssmat50\n",
      "Tetrapoda fragpair_gapless_lf0_rf0_edssmat50\n",
      "Tetrapoda frag_pairwise_gapless_embedding_lf5_rf5\n",
      "Tetrapoda frag_pairwise_gapless_embedding_lf0_rf0\n"
     ]
    }
   ],
   "source": [
    "temp = df.loc[0]\n",
    "og = group_tools.ConserGene(temp[\"json_file\"])\n",
    "og.load_levels(filepath_converter=convert_jsonfile_to_relative)\n",
    "# lvlaln=og.get_aln_score_obj('Metazoa', 'aln_shannon_entropy')\n",
    "# for aa, s in zip(lvlaln.hit_aln_sequence, lvlaln.hit_aln_scores):\n",
    "#     print(aa, s)\n",
    "for level, lvlo in og.level_objects.items():\n",
    "    for scorekey in lvlo.conservation_scores:\n",
    "        print(level, scorekey)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## add score lists for a pairwise score and alignment score at a specific phylogenetic level\n",
    "- level - Metazoa\n",
    "- score_keys\n",
    "  - aln_property_entropy\n",
    "  - fragpair_gapless_lf5_rf5_edssmat50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def json_2_z_score_list(json_file, level, scorekey):\n",
    "    og = group_tools.ConserGene(\n",
    "        json_file, filepath_converter=convert_jsonfile_to_relative\n",
    "    )\n",
    "    if level not in og.levels_passing_filters:\n",
    "        return\n",
    "    lvlo = og.get_level_obj(level, filepath_converter=convert_jsonfile_to_relative)\n",
    "    if scorekey not in lvlo.conservation_scores:\n",
    "        return\n",
    "    if \"hit_z_scores\" not in lvlo.conservation_scores[scorekey]:\n",
    "        return\n",
    "    return lvlo.conservation_scores[scorekey][\"hit_z_scores\"]\n",
    "\n",
    "def add_scorelist_2_df(df, level, scorekey):\n",
    "    colname = f\"{level}_{scorekey}_z_scores\"\n",
    "    df[colname] = df[\"json_file\"].apply(\n",
    "        lambda x: json_2_z_score_list(x, level, scorekey)\n",
    "    )\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "from attrs import asdict, define, field, validators\n",
    "\n",
    "\n",
    "@define\n",
    "class PairwiseScoreResults:\n",
    "    flanked_hit: str\n",
    "    flanked_hit_start_position_in_idr: int\n",
    "    original_hit_st_in_flanked_hit: int\n",
    "    original_hit_end_in_flanked_hit: int\n",
    "    score_function_name: str\n",
    "    score_params: dict\n",
    "    lflank: int\n",
    "    rflank: int\n",
    "    matrix_file: str | Path\n",
    "    flanked_hit_sequence: str\n",
    "    flanked_hit_scores: list\n",
    "    flanked_hit_z_scores: list\n",
    "    hit_sequence: str\n",
    "    hit_scores: list\n",
    "    hit_z_scores: list\n",
    "    mat2score_params: dict\n",
    "\n",
    "    def __attrs_post_init__(self):\n",
    "        self.matrix_file = convert_jsonfile_to_relative(self.matrix_file)\n",
    "\n",
    "\n",
    "@dataclass\n",
    "class AlnScoreResults:\n",
    "    file: str\n",
    "    score_function_name: str\n",
    "    score_params: dict\n",
    "    hit_scores: list\n",
    "    hit_z_scores: list\n",
    "\n",
    "\n",
    "def slice_aln_scores(lvlo: group_tools.LevelAlnScore, aln_start, aln_end):\n",
    "    hit_slice = slice(aln_start, aln_end + 1)\n",
    "    hit_scores = lvlo.scores[hit_slice]\n",
    "    hit_z_scores = lvlo.z_scores[hit_slice]\n",
    "    hit_aln_seq = lvlo.query_aln_sequence[hit_slice]\n",
    "    return hit_scores, hit_z_scores, hit_aln_seq\n",
    "\n",
    "\n",
    "def json2logoplot_alnscore(\n",
    "    jsonfile, score_key, with_gaps=False, axes=None, level=\"Vertebrata\", flank=5\n",
    "):\n",
    "    og = group_tools.ConserGene(\n",
    "        jsonfile, filepath_converter=convert_jsonfile_to_relative\n",
    "    )\n",
    "    lvlo = og.get_aln_score_obj(\n",
    "        level, score_key, filepath_converter=convert_jsonfile_to_relative\n",
    "    )\n",
    "    # aln = jch_alignment.jch_alignment(lvlo.aln, og.query_gene_id)\n",
    "    flst, flend, flhit = tools.pad_hit(\n",
    "        og.query_idr_sequence,\n",
    "        og.hit_st_in_idr,\n",
    "        og.hit_end_in_idr,\n",
    "        l_flank=flank,\n",
    "        r_flank=flank,\n",
    "    )\n",
    "    query_idr, index = tools.reindex_alignment_str(\n",
    "        lvlo.query_aln_sequence[lvlo.idr_aln_start : lvlo.idr_aln_end + 1]\n",
    "    )\n",
    "    flstaln, flendaln = index[flst], index[flend]\n",
    "    flanked_hit_scores, flanked_hit_z_scores, flhit_aln_seq = slice_aln_scores(\n",
    "        lvlo, flstaln + lvlo.idr_aln_start, flendaln + lvlo.idr_aln_start\n",
    "    )\n",
    "    idr_aln = lvlo.aln[:, lvlo.idr_aln_start : lvlo.idr_aln_end + 1]\n",
    "    flhit_aln = idr_aln[:, flstaln : flendaln + 1]\n",
    "\n",
    "    if not with_gaps:\n",
    "        seqlist, query_slice, nongapinds = score_plots.strip_gaps_from_slice(\n",
    "            flhit_aln, flhit_aln_seq\n",
    "        )\n",
    "        score_list = list(np.array(flanked_hit_z_scores)[nongapinds])\n",
    "    else:\n",
    "        seqlist = [str(i.seq) for i in list(flhit_aln)]\n",
    "        query_slice = flhit_aln_seq\n",
    "        score_list = flanked_hit_z_scores\n",
    "    if axes is None:\n",
    "        fig, axes = plt.subplots(nrows=2, ncols=1, figsize=(10, 4))\n",
    "    score_plots.plot_score_bar_plot(\n",
    "        ax=axes[0],\n",
    "        score_list=score_list,\n",
    "        query_seq=query_slice,\n",
    "    )\n",
    "    score_plots.plot_logo(\n",
    "        ax=axes[1],\n",
    "        str_list=seqlist,\n",
    "        tick_label_str=query_slice,\n",
    "    )\n",
    "    counts = pssms.alignment_2_counts(seqlist, show_plot=False, heatmap=False)\n",
    "    return counts\n",
    "\n",
    "\n",
    "def json2logoplot(\n",
    "    jsonfile, score_key, rbm: bool = False, axes=None, level=\"Vertebrata\"\n",
    "):\n",
    "    og = group_tools.ConserGene(\n",
    "        jsonfile, filepath_converter=convert_jsonfile_to_relative\n",
    "    )\n",
    "    lvlo = og.get_level_obj(level, filepath_converter=convert_jsonfile_to_relative)\n",
    "    result = PairwiseScoreResults(**lvlo.conservation_scores[score_key])\n",
    "    mat_dict = pairwise_tools.import_pairwise_matrices(result.matrix_file)\n",
    "    subseqdf = mat_dict[\"subseq_dataframe\"]\n",
    "    subseqdf = subseqdf.fillna(\"-\" * len(result.flanked_hit))\n",
    "    if rbm:\n",
    "        rbmdf = mat_dict[\"reciprocal_best_match_dataframe\"]\n",
    "        rbmdf = rbmdf.fillna(False)\n",
    "        hitdf = pd.concat(\n",
    "            [\n",
    "                subseqdf.loc[result.flanked_hit_start_position_in_idr],\n",
    "                rbmdf.loc[result.flanked_hit_start_position_in_idr],\n",
    "            ],\n",
    "            axis=1,\n",
    "            keys=[\"subseq\", \"rbm\"],\n",
    "        )\n",
    "        hitdf.loc[\"reference_kmer\", \"rbm\"] = True\n",
    "        seqlist = hitdf[hitdf[\"rbm\"]][\"subseq\"].to_list()\n",
    "    else:\n",
    "        seqlist = subseqdf.loc[result.flanked_hit_start_position_in_idr].to_list()\n",
    "    if axes is None:\n",
    "        fig, axes = plt.subplots(nrows=2, ncols=1, figsize=(10, 4))\n",
    "    if not rbm:\n",
    "        score_plots.plot_score_bar_plot(\n",
    "            ax=axes[0],\n",
    "            score_list=result.flanked_hit_z_scores,\n",
    "            query_seq=result.flanked_hit_sequence,\n",
    "        )\n",
    "    score_plots.plot_logo(\n",
    "        ax=axes[1], str_list=seqlist, tick_label_str=result.flanked_hit_sequence\n",
    "    )\n",
    "    counts = pssms.alignment_2_counts(seqlist, show_plot=False, heatmap=False)\n",
    "    return counts, result\n",
    "\n",
    "\n",
    "def plots_from_df(df, level, pairkey = \"fragpair_gapless_lf5_rf5_edssmat50\", output_folder=None, flank=5):\n",
    "    counter = 0\n",
    "    for i, row in df.iterrows():\n",
    "        jsonfile = row[\"json_file\"]\n",
    "        fig, ax = plt.subplots(nrows=2, ncols=3, figsize=(20, 4))\n",
    "        # _=json2logoplot(jsonfile, 'fragpair_gapless_lf5_rf5_edssmat50', rbm=True, axes=ax[:,0], level=level)\n",
    "        # _=json2logoplot(jsonfile, 'frag_pairwise_gapless_embedding_lf5_rf5', rbm=False, axes=ax[:,1])\n",
    "        _ = json2logoplot_alnscore(\n",
    "            jsonfile,\n",
    "            \"aln_property_entropy\",\n",
    "            axes=ax[:, 0],\n",
    "            level=level,\n",
    "            with_gaps=True,\n",
    "            flank=flank,\n",
    "        )\n",
    "        # automatically uses whatever scoring parameters are in the json files (should be just normal right now)\n",
    "        _ = json2logoplot_alnscore(\n",
    "            jsonfile,\n",
    "            \"aln_property_entropy\",\n",
    "            axes=ax[:, 1],\n",
    "            level=level,\n",
    "            with_gaps=False,\n",
    "            flank=flank,\n",
    "        )\n",
    "        _ = json2logoplot(\n",
    "            jsonfile,\n",
    "            pairkey,\n",
    "            rbm=False,\n",
    "            axes=ax[:, 2],\n",
    "            level=level,\n",
    "        )\n",
    "        for axi in ax[0, :]:\n",
    "            axi.set_ylim([-4, 4])\n",
    "        # set the fontsize\n",
    "        # _, result=json2logoplot(jsonfile, 'fragpair_gapless_lf5_rf5_edssmat50', rbm=True, axes=ax[:,2], level=level)\n",
    "        # score_plots.plot_score_bar_plot(\n",
    "        #     ax=ax[0,2],\n",
    "        #     score_list=[0]*5+row['score_list_fragpair_gapless_lf5_rf5_edssmat50'],# this will be the RBM score list\n",
    "        #     query_seq=result.flanked_hit_sequence,\n",
    "        # )\n",
    "        # set fig title\n",
    "        # fig.suptitle(f'{row[\"name\"]}-{row[\"UniprotID\"]}-{row[\"reference_index\"]}')\n",
    "        plt.tight_layout()\n",
    "        if output_folder is not None:\n",
    "            # output_folder2 = Path(output_folder) / f'{row[\"reference_index\"]}'\n",
    "            # output_folder2.mkdir(parents=True, exist_ok=True)\n",
    "            # fig.savefig(\n",
    "            #     output_folder2\n",
    "            #     / f'{level}-{row[\"name\"]}-{row[\"UniprotID\"]}-{row[\"pair-aln\"]:.2f}-{row[\"gene_id\"]}.png',\n",
    "            #     bbox_inches=\"tight\",\n",
    "            #     dpi=300,\n",
    "            # )\n",
    "            fig.savefig(\n",
    "                output_folder\n",
    "                / f'{counter}-{row[\"reference_index\"]}-{row[\"name\"]}-{row[\"UniprotID\"]}-{row[\"pair-aln\"]:.2f}-{row[\"gene_id\"]}.png',\n",
    "                bbox_inches=\"tight\",\n",
    "                dpi=300,\n",
    "            )\n",
    "            plt.close(fig)\n",
    "            counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We lost a lot of motifs mostly because there weren't enough points in the z-score backgrounds. there are either too many gaps in the alignments or not enough k-mers in the pairwise method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "LEVEL = \"Metazoa\"\n",
    "dftemp = df.copy()\n",
    "dftemp = add_scorelist_2_df(dftemp, LEVEL, \"aln_property_entropy\")\n",
    "dftemp = add_scorelist_2_df(dftemp, LEVEL, \"fragpair_gapless_lf5_rf5_edssmat50\")\n",
    "df2 = dftemp[\n",
    "    ~dftemp[\n",
    "        [\n",
    "            f\"{LEVEL}_aln_property_entropy_z_scores\",\n",
    "            f\"{LEVEL}_fragpair_gapless_lf5_rf5_edssmat50_z_scores\",\n",
    "        ]\n",
    "    ]\n",
    "    .isna()\n",
    "    .any(axis=1)\n",
    "].copy()\n",
    "cols = [i for i in df2.columns if \"_scores\" in i]\n",
    "df2[\"weight_array\"] = df2[\"ELM_motif_class\"].map(position_weights)\n",
    "df2 = df2.drop(\"odb_seq\", axis=1)\n",
    "df2[\"weight_array\"].value_counts()\n",
    "# df2['weight_array'].isna().sum()\n",
    "for col in cols:\n",
    "    df2[col + \"_weighted_mean\"] = df2.apply(\n",
    "        lambda x: np.average(x[col], weights=x[\"weight_array\"]), axis=1\n",
    "    )\n",
    "cols = [i for i in df2.columns if \"weighted_mean\" in i]\n",
    "rn = {\n",
    "    f\"{LEVEL}_aln_property_entropy_z_scores_weighted_mean\": \"aln\",\n",
    "    f\"{LEVEL}_fragpair_gapless_lf5_rf5_edssmat50_z_scores_weighted_mean\": \"pair\",\n",
    "}\n",
    "df2 = df2.rename(columns=rn)\n",
    "df2[\"pair-aln\"] = df2[\"pair\"] - df2[\"aln\"]\n",
    "enadf = df2[df2[\"ELM_motif_class\"] == \"enah_LPPPP_FPPPP\"].copy()\n",
    "df2.to_csv(\"score_table.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "35\n"
     ]
    }
   ],
   "source": [
    "output_folder = Path(f\"./plots_ena/\")\n",
    "output_folder.mkdir(exist_ok=True)\n",
    "temp = enadf[enadf[\"pair-aln\"] > 2].copy()\n",
    "print(len(temp))\n",
    "# temp=temp[~temp['UniprotID'].isna()]\n",
    "# temp[temp['verified interaction']].sort_values('pair', ascending=False).head(10)f\n",
    "temp = temp.sort_values(\"pair\", ascending=False)\n",
    "plt.rcParams.update({\"font.size\": 14})\n",
    "plots_from_df(temp, \"Metazoa\", output_folder=output_folder, flank=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "LEVEL = \"Vertebrata\"\n",
    "dftemp = df.copy()\n",
    "dftemp = add_scorelist_2_df(dftemp, LEVEL, \"aln_property_entropy\")\n",
    "dftemp = add_scorelist_2_df(dftemp, LEVEL, \"fragpair_gapless_lf5_rf5_edssmat50\")\n",
    "df2 = dftemp[\n",
    "    ~dftemp[\n",
    "        [\n",
    "            f\"{LEVEL}_aln_property_entropy_z_scores\",\n",
    "            f\"{LEVEL}_fragpair_gapless_lf5_rf5_edssmat50_z_scores\",\n",
    "        ]\n",
    "    ]\n",
    "    .isna()\n",
    "    .any(axis=1)\n",
    "].copy()\n",
    "cols = [i for i in df2.columns if \"_scores\" in i]\n",
    "df2[\"weight_array\"] = df2[\"ELM_motif_class\"].map(position_weights)\n",
    "df2 = df2.drop(\"odb_seq\", axis=1)\n",
    "df2[\"weight_array\"].value_counts()\n",
    "# df2['weight_array'].isna().sum()\n",
    "df2.to_csv(\"score_table.csv\", index=False)\n",
    "for col in cols:\n",
    "    df2[col + \"_weighted_mean\"] = df2.apply(\n",
    "        lambda x: np.average(x[col], weights=x[\"weight_array\"]), axis=1\n",
    "    )\n",
    "cols = [i for i in df2.columns if \"weighted_mean\" in i]\n",
    "rn = {\n",
    "    f\"{LEVEL}_aln_property_entropy_z_scores_weighted_mean\": \"aln\",\n",
    "    f\"{LEVEL}_fragpair_gapless_lf5_rf5_edssmat50_z_scores_weighted_mean\": \"pair\",\n",
    "}\n",
    "df2 = df2.rename(columns=rn)\n",
    "df2[\"pair-aln\"] = df2[\"pair\"] - df2[\"aln\"]\n",
    "enadf = df2[df2[\"ELM_motif_class\"] == \"enah_LPPPP_FPPPP\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    }
   ],
   "source": [
    "output_folder = Path(f\"./plots_ena_Vertebrata/\")\n",
    "output_folder.mkdir(exist_ok=True)\n",
    "temp = enadf[enadf[\"pair-aln\"] > 1].copy()\n",
    "temp = temp[temp[\"pair\"] > 0.5].copy()\n",
    "print(len(temp))\n",
    "# temp=temp[~temp['UniprotID'].isna()]\n",
    "# temp[temp['verified interaction']].sort_values('pair', ascending=False).head(10)f\n",
    "temp = temp.sort_values(\"pair\", ascending=False)\n",
    "# plt.rcParams.update({\"font.size\": 14})\n",
    "plots_from_df(temp, \"Vertebrata\", output_folder=output_folder, flank=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Metazoa_aln_property_entropy_z_scores',\n",
       " 'Metazoa_fragpair_gapless_lf5_rf5_edssmat50_z_scores',\n",
       " 'Vertebrata_aln_property_entropy_z_scores',\n",
       " 'Vertebrata_fragpair_gapless_lf5_rf5_edssmat50_z_scores']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference_index</th>\n",
       "      <th>ELM_motif_class</th>\n",
       "      <th>Organism</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>regex</th>\n",
       "      <th>hit_sequence</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>hit start position</th>\n",
       "      <th>hit end position</th>\n",
       "      <th>verified interaction</th>\n",
       "      <th>name</th>\n",
       "      <th>json_file</th>\n",
       "      <th>critical_error</th>\n",
       "      <th>flanked_hit</th>\n",
       "      <th>Metazoa_aln_property_entropy_z_scores</th>\n",
       "      <th>Metazoa_fragpair_gapless_lf5_rf5_edssmat50_z_scores</th>\n",
       "      <th>Vertebrata_aln_property_entropy_z_scores</th>\n",
       "      <th>Vertebrata_fragpair_gapless_lf5_rf5_edssmat50_z_scores</th>\n",
       "      <th>weight_array</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>O08838</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:004cdf</td>\n",
       "      <td>356</td>\n",
       "      <td>358</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>LDLDFDPFKPDVT</td>\n",
       "      <td>[-0.7849110575078907, -1.1470965302653118, -0....</td>\n",
       "      <td>[1.2646695131128232, -0.6230319877472846, 0.10...</td>\n",
       "      <td>[1.2785820647214632, 0.03832446171214316, 1.27...</td>\n",
       "      <td>[1.7302602704326209, 0.30070041842870615, 1.54...</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>399</td>\n",
       "      <td>401</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EAPISDPFAPEPS</td>\n",
       "      <td>[-0.3065326707116909, -1.5818396760602311, -1....</td>\n",
       "      <td>[0.8068631016990869, 0.6504573818486616, 1.619...</td>\n",
       "      <td>[-0.3757825927989007, -1.2417344591634956, -1....</td>\n",
       "      <td>[0.22488942726408537, 0.02772335993562546, -0....</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>473</td>\n",
       "      <td>475</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ACSGNDPFAPSEG</td>\n",
       "      <td>[0.017487595566555075, 1.2401414924355165, 1.2...</td>\n",
       "      <td>[0.306919128843599, 0.5543636025383214, 2.1653...</td>\n",
       "      <td>[1.0671007661976506, 0.88663452426079, 1.06710...</td>\n",
       "      <td>[1.638242435956166, 1.3967010985501278, 1.4820...</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>292</td>\n",
       "      <td>294</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PTPNPDPFRDDPF</td>\n",
       "      <td>[0.6769662402228052, 0.372685432141188, 1.7414...</td>\n",
       "      <td>[3.662421085771397, 3.191840346290177, 4.11101...</td>\n",
       "      <td>[1.2045098328535813, 1.6961993321774382, 1.631...</td>\n",
       "      <td>[1.7308183030776918, 2.0581029493236827, 2.202...</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>297</td>\n",
       "      <td>299</td>\n",
       "      <td>True</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>DPFRDDPFTQPDQ</td>\n",
       "      <td>[-0.48778871074170704, 0.5659504569320113, 1.3...</td>\n",
       "      <td>[2.7705513589276904, 1.0505947964105804, 2.262...</td>\n",
       "      <td>[1.4522314930727687, 1.5176974743943668, 0.480...</td>\n",
       "      <td>[2.074211827645784, 1.8577131695222864, 0.8742...</td>\n",
       "      <td>[1, 1, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2882</th>\n",
       "      <td>2882</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A0A384MDR3</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>IKKPAEDEW</td>\n",
       "      <td>9606_0:004627</td>\n",
       "      <td>81</td>\n",
       "      <td>89</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ALFQDIKKPAEDEWGKTPD</td>\n",
       "      <td>[1.6241450780345585, -1.4089095965466998, -1.0...</td>\n",
       "      <td>[0.8050348655032681, -1.0921869671063063, -0.5...</td>\n",
       "      <td>[1.154489476836621, -1.007434199238479, 1.1817...</td>\n",
       "      <td>[0.5451229600801136, -0.8964587731723821, 1.16...</td>\n",
       "      <td>[0, 0, 0, 1, 0, 1, 0, 0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2883</th>\n",
       "      <td>2883</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Q14209</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>SLVPLEATD</td>\n",
       "      <td>9606_0:00055f</td>\n",
       "      <td>363</td>\n",
       "      <td>371</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>APPPPSLVPLEATDSLLEL</td>\n",
       "      <td>[-1.180650083469508, -1.331294073817295, -1.46...</td>\n",
       "      <td>[-1.0958606737149268, -0.04076692679728417, -0...</td>\n",
       "      <td>[-1.048000273450755, -1.057049107954081, -1.12...</td>\n",
       "      <td>[-1.0663432621108706, -0.29635935974563504, -0...</td>\n",
       "      <td>[0, 0, 0, 1, 0, 1, 0, 0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2885</th>\n",
       "      <td>2885</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>Q9H816</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>FESPEESAD</td>\n",
       "      <td>9606_0:000204</td>\n",
       "      <td>353</td>\n",
       "      <td>361</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>TQGVVFESPEESADQSQAD</td>\n",
       "      <td>None</td>\n",
       "      <td>[3.644601901565253, 3.0552242871772446, 2.8860...</td>\n",
       "      <td>[2.2689217162116884, 2.06794344628975, 1.70054...</td>\n",
       "      <td>[3.952637944651848, 4.207351227077069, 3.17172...</td>\n",
       "      <td>[0, 0, 0, 1, 0, 1, 0, 0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2886</th>\n",
       "      <td>2886</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A4PB67</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>FHQPEEEIE</td>\n",
       "      <td>9606_0:00057e</td>\n",
       "      <td>215</td>\n",
       "      <td>223</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PLRTLFHQPEEEIEDGGLF</td>\n",
       "      <td>[-1.7556914529906844, -1.7727575922158432, -1....</td>\n",
       "      <td>[-0.6420363343827739, -1.1415426882540252, -1....</td>\n",
       "      <td>[-2.2478124490435225, -2.287941325460294, -2.0...</td>\n",
       "      <td>[-0.9342903186837597, -1.1217520698203998, -1....</td>\n",
       "      <td>[0, 0, 0, 1, 0, 1, 0, 0, 1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2887</th>\n",
       "      <td>2887</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>A6NGG8</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>ENSPKETQE</td>\n",
       "      <td>9606_0:00094d</td>\n",
       "      <td>855</td>\n",
       "      <td>863</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>../../benchmark/benchmark_v3/p3_conservation/c...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SSKSTENSPKETQEPGPGE</td>\n",
       "      <td>[-0.21004202158033639, -0.7622876614597881, 0....</td>\n",
       "      <td>[-0.9799376966198027, -0.9519405387678506, 0.0...</td>\n",
       "      <td>[-0.21004202158033639, -0.7622876614597881, 0....</td>\n",
       "      <td>[-0.9799376966198027, -0.9519405387678506, 0.0...</td>\n",
       "      <td>[0, 0, 0, 1, 0, 1, 0, 0, 1]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2152 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      reference_index ELM_motif_class           Organism   UniprotID  \\\n",
       "0                   0  LIG_AP2alpha_2  Rattus norvegicus      O08838   \n",
       "1                   1  LIG_AP2alpha_2  Rattus norvegicus      Q05140   \n",
       "2                   2  LIG_AP2alpha_2  Rattus norvegicus      Q05140   \n",
       "3                   3  LIG_AP2alpha_2       Homo sapiens      P98082   \n",
       "4                   4  LIG_AP2alpha_2       Homo sapiens      P98082   \n",
       "...               ...             ...                ...         ...   \n",
       "2882             2882           TRAF6       Homo sapiens  A0A384MDR3   \n",
       "2883             2883           TRAF6       Homo sapiens      Q14209   \n",
       "2885             2885           TRAF6       Homo sapiens      Q9H816   \n",
       "2886             2886           TRAF6       Homo sapiens      A4PB67   \n",
       "2887             2887           TRAF6       Homo sapiens      A6NGG8   \n",
       "\n",
       "                regex hit_sequence         gene_id  hit start position  \\\n",
       "0              DP[FW]          DPF  10116_0:004cdf                 356   \n",
       "1              DP[FW]          DPF  10116_0:002e5e                 399   \n",
       "2              DP[FW]          DPF  10116_0:002e5e                 473   \n",
       "3              DP[FW]          DPF   9606_0:0016b2                 292   \n",
       "4              DP[FW]          DPF   9606_0:0016b2                 297   \n",
       "...               ...          ...             ...                 ...   \n",
       "2882  ...P.E..[FYWDE]    IKKPAEDEW   9606_0:004627                  81   \n",
       "2883  ...P.E..[FYWDE]    SLVPLEATD   9606_0:00055f                 363   \n",
       "2885  ...P.E..[FYWDE]    FESPEESAD   9606_0:000204                 353   \n",
       "2886  ...P.E..[FYWDE]    FHQPEEEIE   9606_0:00057e                 215   \n",
       "2887  ...P.E..[FYWDE]    ENSPKETQE   9606_0:00094d                 855   \n",
       "\n",
       "      hit end position  verified interaction name  \\\n",
       "0                  358                  True  NaN   \n",
       "1                  401                  True  NaN   \n",
       "2                  475                  True  NaN   \n",
       "3                  294                  True  NaN   \n",
       "4                  299                  True  NaN   \n",
       "...                ...                   ...  ...   \n",
       "2882                89                 False  NaN   \n",
       "2883               371                 False  NaN   \n",
       "2885               361                 False  NaN   \n",
       "2886               223                 False  NaN   \n",
       "2887               863                 False  NaN   \n",
       "\n",
       "                                              json_file critical_error  \\\n",
       "0     ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "1     ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "2     ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "3     ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "4     ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "...                                                 ...            ...   \n",
       "2882  ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "2883  ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "2885  ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "2886  ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "2887  ../../benchmark/benchmark_v3/p3_conservation/c...            NaN   \n",
       "\n",
       "              flanked_hit              Metazoa_aln_property_entropy_z_scores  \\\n",
       "0           LDLDFDPFKPDVT  [-0.7849110575078907, -1.1470965302653118, -0....   \n",
       "1           EAPISDPFAPEPS  [-0.3065326707116909, -1.5818396760602311, -1....   \n",
       "2           ACSGNDPFAPSEG  [0.017487595566555075, 1.2401414924355165, 1.2...   \n",
       "3           PTPNPDPFRDDPF  [0.6769662402228052, 0.372685432141188, 1.7414...   \n",
       "4           DPFRDDPFTQPDQ  [-0.48778871074170704, 0.5659504569320113, 1.3...   \n",
       "...                   ...                                                ...   \n",
       "2882  ALFQDIKKPAEDEWGKTPD  [1.6241450780345585, -1.4089095965466998, -1.0...   \n",
       "2883  APPPPSLVPLEATDSLLEL  [-1.180650083469508, -1.331294073817295, -1.46...   \n",
       "2885  TQGVVFESPEESADQSQAD                                               None   \n",
       "2886  PLRTLFHQPEEEIEDGGLF  [-1.7556914529906844, -1.7727575922158432, -1....   \n",
       "2887  SSKSTENSPKETQEPGPGE  [-0.21004202158033639, -0.7622876614597881, 0....   \n",
       "\n",
       "     Metazoa_fragpair_gapless_lf5_rf5_edssmat50_z_scores  \\\n",
       "0     [1.2646695131128232, -0.6230319877472846, 0.10...    \n",
       "1     [0.8068631016990869, 0.6504573818486616, 1.619...    \n",
       "2     [0.306919128843599, 0.5543636025383214, 2.1653...    \n",
       "3     [3.662421085771397, 3.191840346290177, 4.11101...    \n",
       "4     [2.7705513589276904, 1.0505947964105804, 2.262...    \n",
       "...                                                 ...    \n",
       "2882  [0.8050348655032681, -1.0921869671063063, -0.5...    \n",
       "2883  [-1.0958606737149268, -0.04076692679728417, -0...    \n",
       "2885  [3.644601901565253, 3.0552242871772446, 2.8860...    \n",
       "2886  [-0.6420363343827739, -1.1415426882540252, -1....    \n",
       "2887  [-0.9799376966198027, -0.9519405387678506, 0.0...    \n",
       "\n",
       "               Vertebrata_aln_property_entropy_z_scores  \\\n",
       "0     [1.2785820647214632, 0.03832446171214316, 1.27...   \n",
       "1     [-0.3757825927989007, -1.2417344591634956, -1....   \n",
       "2     [1.0671007661976506, 0.88663452426079, 1.06710...   \n",
       "3     [1.2045098328535813, 1.6961993321774382, 1.631...   \n",
       "4     [1.4522314930727687, 1.5176974743943668, 0.480...   \n",
       "...                                                 ...   \n",
       "2882  [1.154489476836621, -1.007434199238479, 1.1817...   \n",
       "2883  [-1.048000273450755, -1.057049107954081, -1.12...   \n",
       "2885  [2.2689217162116884, 2.06794344628975, 1.70054...   \n",
       "2886  [-2.2478124490435225, -2.287941325460294, -2.0...   \n",
       "2887  [-0.21004202158033639, -0.7622876614597881, 0....   \n",
       "\n",
       "     Vertebrata_fragpair_gapless_lf5_rf5_edssmat50_z_scores  \\\n",
       "0     [1.7302602704326209, 0.30070041842870615, 1.54...       \n",
       "1     [0.22488942726408537, 0.02772335993562546, -0....       \n",
       "2     [1.638242435956166, 1.3967010985501278, 1.4820...       \n",
       "3     [1.7308183030776918, 2.0581029493236827, 2.202...       \n",
       "4     [2.074211827645784, 1.8577131695222864, 0.8742...       \n",
       "...                                                 ...       \n",
       "2882  [0.5451229600801136, -0.8964587731723821, 1.16...       \n",
       "2883  [-1.0663432621108706, -0.29635935974563504, -0...       \n",
       "2885  [3.952637944651848, 4.207351227077069, 3.17172...       \n",
       "2886  [-0.9342903186837597, -1.1217520698203998, -1....       \n",
       "2887  [-0.9799376966198027, -0.9519405387678506, 0.0...       \n",
       "\n",
       "                     weight_array  \n",
       "0                       [1, 1, 1]  \n",
       "1                       [1, 1, 1]  \n",
       "2                       [1, 1, 1]  \n",
       "3                       [1, 1, 1]  \n",
       "4                       [1, 1, 1]  \n",
       "...                           ...  \n",
       "2882  [0, 0, 0, 1, 0, 1, 0, 0, 1]  \n",
       "2883  [0, 0, 0, 1, 0, 1, 0, 0, 1]  \n",
       "2885  [0, 0, 0, 1, 0, 1, 0, 0, 1]  \n",
       "2886  [0, 0, 0, 1, 0, 1, 0, 0, 1]  \n",
       "2887  [0, 0, 0, 1, 0, 1, 0, 0, 1]  \n",
       "\n",
       "[2152 rows x 19 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1dfc83f8",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "traf = df2[df2[\"ELM_motif_class\"] == \"TRAF6\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce834534",
   "metadata": {
    "lines_to_next_cell": 2
   },
   "outputs": [],
   "source": [
    "\n",
    "output_folder = Path(f\"./plots_{LEVEL}_traf\")\n",
    "output_folder.mkdir(exist_ok=True)\n",
    "temp = traf[traf[\"pair-aln\"] > 1].copy()\n",
    "temp = temp[temp[\"pair\"] > 1].copy()\n",
    "temp = temp.sort_values(\"pair\", ascending=False)\n",
    "plots_from_df(temp, \"Metazoa\", output_folder=output_folder, pairkey=\"fragpair_gapless_lf0_rf0_edssmat50\", flank=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "46cf5c24",
   "metadata": {},
   "source": [
    "Q9Y6W5\n",
    "WASF2_HUMAN\n",
    "Part of the WAVE complex that regulates lamellipodia formation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plots_from_df(\n",
    "    temp[~temp[\"verified interaction\"]]\n",
    "    .sort_values(\"mean_score_fragpair_gapless_lf5_rf5_edssmat50\", ascending=False)\n",
    "    .head(10),\n",
    "    \"Metazoa\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vert_df.sort_values(\"aln-pair\", ascending=False).head(10)\n",
    "temp = vert_df[vert_df[\"aln-pair\"] > 2].copy()\n",
    "temp = temp[~temp[\"UniprotID\"].isna()]\n",
    "plt.rcParams.update({\"font.size\": 16})\n",
    "plots_from_df(\n",
    "    temp[temp[\"verified interaction\"]]\n",
    "    .sort_values(\"mean_score_fragpair_gapless_lf5_rf5_edssmat50\", ascending=False)\n",
    "    .head(10),\n",
    "    \"Vertebrata\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "temp = vert_df[vert_df[\"reference_index\"] == 2243].copy()\n",
    "fig, ax = plt.subplots(nrows=2, figsize=(8, 4))\n",
    "_ = json2logoplot_alnscore(\n",
    "    temp[\"json_file\"].values[0],\n",
    "    \"aln_property_entropy\",\n",
    "    axes=ax,\n",
    "    level=\"Vertebrata\",\n",
    "    with_gaps=True,\n",
    "    flank=0,\n",
    ")\n",
    "ax[0].set_ylim([-3, 3])\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "df2.loc[2887].to_json(\"./test.json\", orient=\"records\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.8155002336777777"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.average(\n",
    "    [\n",
    "        -0.2100420216,\n",
    "        -0.7622876615,\n",
    "        0.2637430374,\n",
    "        -0.3815422762,\n",
    "        -0.9570456609,\n",
    "        -1.3029297428,\n",
    "        -1.2251521697,\n",
    "        -1.4368219208,\n",
    "        -1.327423687,\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.0039652353333333"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.average(\n",
    "    [\n",
    "        -0.2100420216,\n",
    "        -0.7622876615,\n",
    "        0.2637430374,\n",
    "        -0.3815422762,\n",
    "        -0.9570456609,\n",
    "        -1.3029297428,\n",
    "        -1.2251521697,\n",
    "        -1.4368219208,\n",
    "        -1.327423687,\n",
    "    ],\n",
    "    weights=[0, 0, 0, 1, 0, 1, 0, 0, 1],\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-1.0039652353333333"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.average([-0.3815422762, -1.3029297428, -1.327423687])\n",
    "np.mean([-0.3815422762, -1.3029297428, -1.327423687])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41aaf0f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "jupytext": {
   "formats": "ipynb,py:percent"
  },
  "kernelspec": {
   "display_name": "odb_conservation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
