{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This should be run only after `../../ELM_instances/build_instance_map.ipynb` and `../../all_human_odb_sequences_removed_duplicates/get_all_human_proteins_in_odb.ipynb` have been run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# run with [pipeline 1](https://github.com/jacksonh1/orthogroup_generation) environment\n",
    "`mamba activate odb_groups_x86`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "add path to the IUPRED2A library in the file `./iupred_tools.py` as the variable `IUPRED_PATH`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from pyprojroot import here\n",
    "\n",
    "import local_orthoDB_group_pipeline.sql_queries as sql_queries\n",
    "import local_seqtools.general_utils as tools\n",
    "import iupred_tools as iuptools\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import yaml\n",
    "from Bio import Align, AlignIO, Seq, SeqIO\n",
    "# from local_env_variables import project_filepaths as fp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "BENCHMARK_DIR = here() / \"benchmark\" / \"benchmark_v3\" / \"p1_table\"\n",
    "BENCHMARK_DIR.mkdir(parents=True, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Verified interactions/regexes from the ELM database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def in_any_idrs(idr_list: list, start: int, end: int):\n",
    "    for idr in idr_list:\n",
    "        if idr[0] <= start and end <= idr[1]:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "\n",
    "def regex_search(regex, seqrecord_list, idr_map: dict = None):\n",
    "    odb_id_ex = []\n",
    "    hit_start_pos_ex = []\n",
    "    hit_end_pos_ex = []\n",
    "    motif_match_ex = []\n",
    "\n",
    "    for record in seqrecord_list:\n",
    "        sequence = str(record.seq)\n",
    "        if idr_map is not None:\n",
    "            idrs = idr_map[record.id]\n",
    "        for m in tools.get_regex_matches(regex, sequence):\n",
    "            start_pos = m[1]\n",
    "            end_pos = m[2]\n",
    "            if idr_map is not None:\n",
    "                if len(idrs) == 0:\n",
    "                    continue\n",
    "                if not in_any_idrs(idrs, start_pos, end_pos):\n",
    "                    continue\n",
    "            odb_id_ex.append(record.id)\n",
    "            hit_start_pos_ex.append(start_pos)\n",
    "            hit_end_pos_ex.append(end_pos)\n",
    "            motif_match_ex.append(m[0])\n",
    "\n",
    "    df_fp = pd.DataFrame(\n",
    "        {\n",
    "            \"odb_id\": odb_id_ex,\n",
    "            \"odb_mot_st\": hit_start_pos_ex,\n",
    "            \"odb_mot_end\": hit_end_pos_ex,\n",
    "            \"motif_match\": motif_match_ex,\n",
    "        }\n",
    "    )\n",
    "    df_fp[\"regex\"] = regex\n",
    "    return df_fp\n",
    "\n",
    "\n",
    "def load_verified_from_instances(\n",
    "    motif_class_name,\n",
    "    elm_instances_df,\n",
    "    allowed_organisms=[\"Homo sapiens\", \"Mus musculus\", \"Rattus norvegicus\"],\n",
    "):\n",
    "    \"\"\"Load the verified instances for 1 slim class from the table of verified instances. Filters out instances not in `allowed_organisms`\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    motif_class_name : str\n",
    "        name of the motif class (in the table)\n",
    "    elm_instances_df : pd.Dataframe\n",
    "        Instances from the ELM (should be preprocessed)\n",
    "    allowed_organisms : list, optional\n",
    "        list of organisms to allow, by default [\"Homo sapiens\", \"Mus musculus\", \"Rattus norvegicus\"]\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    tuple\n",
    "        returns the pd.Dataframe of instances and the regular expression of the motif\n",
    "    \"\"\"\n",
    "    print(f\"Getting motif benchmark table for {motif_class_name}\")\n",
    "\n",
    "    # import known binders\n",
    "    df = elm_instances_df.copy()\n",
    "    df = df[df[\"ELMIdentifier\"] == motif_class_name].copy()\n",
    "    df[\"verified interaction\"] = True\n",
    "    df = df.rename(\n",
    "        columns={\n",
    "            \"uniprot2map\": \"UniprotID\",\n",
    "            \"Regex\": \"regex\",\n",
    "            \"ProteinName\": \"name\",\n",
    "            \"hit_sequence\": \"flanked_motif_match\",\n",
    "        }\n",
    "    )\n",
    "    df = df[df[\"InstanceLogic\"] == \"true positive\"]\n",
    "    df2 = df[\n",
    "        [\n",
    "            # \"Accession_inst\",\n",
    "            # \"Accession_class\",\n",
    "            \"Organism\",\n",
    "            # \"Probability\",\n",
    "            # \"ELMIdentifier\",\n",
    "            # \"name\",\n",
    "            \"Primary_Acc\",\n",
    "            \"Accessions\",\n",
    "            # \"PDB\",\n",
    "            # \"FunctionalSiteName\",\n",
    "            # \"Description\",\n",
    "            # 'uniprot2map',\n",
    "            \"UniprotID\",\n",
    "            # \"fasta_id\",\n",
    "            \"regex\",\n",
    "            \"motif_match\",\n",
    "            # \"flanked_motif_match\",\n",
    "            # \"mot_start_pos_in_hit\",\n",
    "            \"odb_id\",\n",
    "            # \"odb_hit_start\",\n",
    "            # \"odb_hit_end\",\n",
    "            \"odb_mot_st\",\n",
    "            \"odb_mot_end\",\n",
    "            \"verified interaction\",\n",
    "        ]\n",
    "    ].copy()\n",
    "\n",
    "    print(\"restrict known binders to a subset of organisms (only ones close to human)\")\n",
    "\n",
    "    print(f\"allowed organisms for known binding interactions {allowed_organisms}\")\n",
    "    df2 = df2[df2[\"Organism\"].isin(allowed_organisms)]\n",
    "    reglist = df2[\"regex\"].unique()\n",
    "    assert len(reglist) == 1, \"More than one regex in dataframe of known binders\"\n",
    "    regex = reglist[0]\n",
    "    return df2, regex\n",
    "\n",
    "\n",
    "def add_clashes_with_known_instances(df_bg_in, instances_df):\n",
    "    df_bg = df_bg_in.copy()\n",
    "    clash_map = {}\n",
    "    for i, row in df_bg.iterrows():\n",
    "        elm_clash_flag = False\n",
    "        if row[\"odb_id\"] in instances_df[\"odb_id\"].values:\n",
    "            potential_overlaps = instances_df[instances_df[\"odb_id\"] == row[\"odb_id\"]]\n",
    "            for j, instance_row in potential_overlaps.iterrows():\n",
    "                if (\n",
    "                    row[\"odb_mot_st\"] <= instance_row[\"odb_mot_end\"]\n",
    "                    and row[\"odb_mot_end\"] >= instance_row[\"odb_mot_st\"]\n",
    "                ):\n",
    "                    elm_clash_flag = True\n",
    "            if elm_clash_flag:\n",
    "                clash_map[i] = True\n",
    "            else:\n",
    "                clash_map[i] = False\n",
    "        else:\n",
    "            clash_map[i] = False\n",
    "    df_bg[\"clash\"] = df_bg.index.map(clash_map)\n",
    "    return df_bg\n",
    "\n",
    "\n",
    "def get_bg_seqs(\n",
    "    bg_seq_fasta_file,\n",
    "    regex,\n",
    "    instance_df,\n",
    "    idr_map: dict = None,\n",
    "    bg_organism=\"Homo sapiens\",\n",
    "):\n",
    "    faimporter = tools.FastaImporter(bg_seq_fasta_file)\n",
    "    bg_search_seqs = faimporter.import_as_list()\n",
    "    df_bg = regex_search(regex, bg_search_seqs, idr_map=idr_map)\n",
    "    df_bg[\"UniprotID\"] = df_bg[\"odb_id\"].apply(sql_queries.odb_gene_id_2_uniprotid)\n",
    "    df_bg[\"verified interaction\"] = False\n",
    "    df_bg[\"Organism\"] = bg_organism\n",
    "    df_bg = add_clashes_with_known_instances(df_bg, instance_df)\n",
    "    df_bg = df_bg[~df_bg[\"clash\"]]\n",
    "    df_bg = df_bg.drop(columns=\"clash\")\n",
    "    return df_bg\n",
    "\n",
    "\n",
    "def merge_verified_and_fp_motifs(df_verified, df_bg, cluster_dict=None):\n",
    "    # remove overlap with verified binders\n",
    "    verified_binder_odids = list(df_verified[\"odb_id\"].unique())\n",
    "    if cluster_dict is not None:\n",
    "        # for each verified binder, add all cluster members so that they're removed from fp as well\n",
    "        for cluster in cluster_dict:\n",
    "            if any([i in verified_binder_odids for i in cluster_dict[cluster]['all_members']]):\n",
    "                print('added cluster members')\n",
    "                verified_binder_odids.extend(cluster_dict[cluster]['all_members'])\n",
    "        verified_binder_odids = list(set(verified_binder_odids))\n",
    "    df_bg = df_bg[~df_bg[\"odb_id\"].isin(verified_binder_odids)].copy()\n",
    "    assert (\n",
    "        df_bg[\"UniprotID\"].isin(df_verified[\"UniprotID\"].unique()).sum() == 0\n",
    "    ), \"False positives overlap with verified binders by uniprot id\"\n",
    "    df_verified[\"regex\"] = df_verified[\"regex\"].astype(str)\n",
    "    df_verified[\"verified interaction\"] = df_verified[\"verified interaction\"].astype(\n",
    "        bool\n",
    "    )\n",
    "\n",
    "    # # merge\n",
    "    dfm = pd.concat([df_verified, df_bg]).reset_index(drop=True)\n",
    "    return dfm\n",
    "    # dfm.to_csv(f\"./{motif_class_name}_motifs.csv\", index=False)\n",
    "    # for col in dfm.columns: print(col)\n",
    "\n",
    "\n",
    "def main(\n",
    "    elm_class,\n",
    "    elm_instances_file,\n",
    "    bg_seq_fasta_file,\n",
    "    idr_map,\n",
    "    bg_organism,\n",
    "    subsample: int|None=None,\n",
    "    cluster_dict=None,\n",
    "):\n",
    "    elm_instances_df = pd.read_csv(elm_instances_file)\n",
    "    dftp, regex = load_verified_from_instances(elm_class, elm_instances_df)\n",
    "    bg_df = get_bg_seqs(\n",
    "        bg_seq_fasta_file,\n",
    "        regex,\n",
    "        elm_instances_df,\n",
    "        idr_map=idr_map,\n",
    "        bg_organism=bg_organism,\n",
    "    )\n",
    "    if subsample is not None and len(bg_df) > subsample:\n",
    "        bg_df = bg_df.sample(subsample, random_state=42)\n",
    "    dfm = merge_verified_and_fp_motifs(dftp, bg_df, cluster_dict=cluster_dict)\n",
    "    dfm[\"ELM_motif_class\"] = elm_class\n",
    "    return dfm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## load idr map for all human proteins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_sequences = \"../../../../data/orthodb_clustered_species_proteins/9606_6/all_human_proteins_in_odb.fasta\"\n",
    "\n",
    "idr_map = {}\n",
    "faimporter = tools.FastaImporter(human_sequences)\n",
    "all_human_seq_dict = faimporter.import_as_dict()\n",
    "for k, v in all_human_seq_dict.items():\n",
    "    idr_map[k] = iuptools.main_find_idr_regions(\n",
    "        str(v.seq),\n",
    "        iupred_cutoff=0.4,\n",
    "        gap_merge_threshold=10,\n",
    "        idr_min_length=8,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## generating tables for a variety of ELM motifs (classes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "ELM_CLASSES_4_BENCHMARK = [\n",
    "    \"LIG_AP2alpha_2\",\n",
    "    \"LIG_EH_1\",\n",
    "    \"LIG_SH2_GRB2like\",\n",
    "    \"LIG_SH3_CIN85_PxpxPR_1\",\n",
    "    # \"LIG_PDZ_Class_1\",\n",
    "    \"DOC_WW_Pin1_4\", # this will have to be subsampled because the regex is way too general\n",
    "    # \"LIG_LIR_Gen_1\",\n",
    "    \"LIG_14-3-3_CanoR_1\"\n",
    "]\n",
    "# This is to set the number of background sequences to sample for each motif class in the cases where there are too many background matches. This happens when the motif regex being too general\n",
    "CLASS_BG_SAMPLING = {\n",
    "    \"LIG_AP2alpha_2\": 300,\n",
    "    \"LIG_EH_1\": 300,\n",
    "    \"LIG_SH2_GRB2like\": 300,\n",
    "    \"LIG_SH3_CIN85_PxpxPR_1\": 300,\n",
    "    # \"LIG_PDZ_Class_1\": 300,\n",
    "    \"DOC_WW_Pin1_4\": 300, # this will have to be subsampled because the regex is way too general\n",
    "    # \"LIG_LIR_Gen_1\": 300,\n",
    "    \"LIG_14-3-3_CanoR_1\": 300\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "BG_SEQ_FASTA_FILE = \"../../../../data/orthodb_clustered_species_proteins/9606_6/all_human_proteins_in_odb_clustered_c0_95.fasta\"\n",
    "BG_CLUSTER_JSON_FILE = \"../../../../data/orthodb_clustered_species_proteins/9606_6/all_human_proteins_in_odb_clustered_c0_95.json\"\n",
    "BG_ORGANISM = \"Homo sapiens\"\n",
    "ELM_INSTANCES_FILE = \"../../../../data/ELM/2024-02-09-ELM_instances/elm_instances_with_hit_sequence.csv\"\n",
    "import json\n",
    "with open(BG_CLUSTER_JSON_FILE) as f:\n",
    "    BG_CLUSTER_DICT = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# elm_instances_df = pd.read_csv(ELM_INSTANCES_FILE)\n",
    "# dftp, regex = load_verified_from_instances(ELM_CLASSES_4_BENCHMARK[0], elm_instances_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# df = get_bg_seqs(BG_SEQ_FASTA_FILE, regex, elm_instances_df, idr_map=idr_map, bg_organism=BG_ORGANISM)\n",
    "# df[df['clash']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting motif benchmark table for LIG_AP2alpha_2\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "Getting motif benchmark table for LIG_EH_1\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "Getting motif benchmark table for LIG_SH2_GRB2like\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "Getting motif benchmark table for LIG_SH3_CIN85_PxpxPR_1\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "Getting motif benchmark table for DOC_WW_Pin1_4\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "Getting motif benchmark table for LIG_14-3-3_CanoR_1\n",
      "restrict known binders to a subset of organisms (only ones close to human)\n",
      "allowed organisms for known binding interactions ['Homo sapiens', 'Mus musculus', 'Rattus norvegicus']\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n"
     ]
    }
   ],
   "source": [
    "df_list = []\n",
    "for motif in ELM_CLASSES_4_BENCHMARK:\n",
    "    if motif in CLASS_BG_SAMPLING:\n",
    "        df = main(\n",
    "            elm_class=motif,\n",
    "            elm_instances_file=ELM_INSTANCES_FILE,\n",
    "            bg_seq_fasta_file=BG_SEQ_FASTA_FILE,\n",
    "            idr_map=idr_map,\n",
    "            bg_organism=BG_ORGANISM,\n",
    "            subsample=CLASS_BG_SAMPLING[motif],\n",
    "            cluster_dict=BG_CLUSTER_DICT\n",
    "        )\n",
    "    else:\n",
    "        df = main(\n",
    "            elm_class=motif,\n",
    "            elm_instances_file=ELM_INSTANCES_FILE,\n",
    "            bg_seq_fasta_file=BG_SEQ_FASTA_FILE,\n",
    "            idr_map=idr_map,\n",
    "            bg_organism=BG_ORGANISM,\n",
    "            cluster_dict=BG_CLUSTER_DICT\n",
    "        )\n",
    "    df_list.append(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Primary_Acc</th>\n",
       "      <th>Accessions</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>regex</th>\n",
       "      <th>motif_match</th>\n",
       "      <th>odb_id</th>\n",
       "      <th>odb_mot_st</th>\n",
       "      <th>odb_mot_end</th>\n",
       "      <th>verified interaction</th>\n",
       "      <th>ELM_motif_class</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>O08838</td>\n",
       "      <td>O08838</td>\n",
       "      <td>O08838</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:004cdf</td>\n",
       "      <td>356</td>\n",
       "      <td>358</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>399</td>\n",
       "      <td>401</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>473</td>\n",
       "      <td>475</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>P98082 A6NES5 Q13598 Q9BTY0 Q9UK04</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>292</td>\n",
       "      <td>294</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>P98082 A6NES5 Q13598 Q9BTY0 Q9UK04</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>297</td>\n",
       "      <td>299</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2087</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O14641</td>\n",
       "      <td>R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...</td>\n",
       "      <td>RVVSWL</td>\n",
       "      <td>9606_0:003d26</td>\n",
       "      <td>83</td>\n",
       "      <td>88</td>\n",
       "      <td>False</td>\n",
       "      <td>LIG_14-3-3_CanoR_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2088</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B7Z8A0</td>\n",
       "      <td>R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...</td>\n",
       "      <td>RINSFE</td>\n",
       "      <td>9606_0:002891</td>\n",
       "      <td>488</td>\n",
       "      <td>493</td>\n",
       "      <td>False</td>\n",
       "      <td>LIG_14-3-3_CanoR_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2089</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q4LE39</td>\n",
       "      <td>R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...</td>\n",
       "      <td>RSIKSETDSTI</td>\n",
       "      <td>9606_0:0000b6</td>\n",
       "      <td>1058</td>\n",
       "      <td>1068</td>\n",
       "      <td>False</td>\n",
       "      <td>LIG_14-3-3_CanoR_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2090</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A8K119</td>\n",
       "      <td>R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...</td>\n",
       "      <td>RSTCNVV</td>\n",
       "      <td>9606_0:002261</td>\n",
       "      <td>246</td>\n",
       "      <td>252</td>\n",
       "      <td>False</td>\n",
       "      <td>LIG_14-3-3_CanoR_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2091</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q6P2M8</td>\n",
       "      <td>R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...</td>\n",
       "      <td>RHSHSGL</td>\n",
       "      <td>9606_0:004e60</td>\n",
       "      <td>411</td>\n",
       "      <td>417</td>\n",
       "      <td>False</td>\n",
       "      <td>LIG_14-3-3_CanoR_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2092 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Organism Primary_Acc                          Accessions  \\\n",
       "0     Rattus norvegicus      O08838                              O08838   \n",
       "1     Rattus norvegicus      Q05140                              Q05140   \n",
       "2     Rattus norvegicus      Q05140                              Q05140   \n",
       "3          Homo sapiens      P98082  P98082 A6NES5 Q13598 Q9BTY0 Q9UK04   \n",
       "4          Homo sapiens      P98082  P98082 A6NES5 Q13598 Q9BTY0 Q9UK04   \n",
       "...                 ...         ...                                 ...   \n",
       "2087       Homo sapiens         NaN                                 NaN   \n",
       "2088       Homo sapiens         NaN                                 NaN   \n",
       "2089       Homo sapiens         NaN                                 NaN   \n",
       "2090       Homo sapiens         NaN                                 NaN   \n",
       "2091       Homo sapiens         NaN                                 NaN   \n",
       "\n",
       "     UniprotID                                              regex  \\\n",
       "0       O08838                                             DP[FW]   \n",
       "1       Q05140                                             DP[FW]   \n",
       "2       Q05140                                             DP[FW]   \n",
       "3       P98082                                             DP[FW]   \n",
       "4       P98082                                             DP[FW]   \n",
       "...        ...                                                ...   \n",
       "2087    O14641  R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...   \n",
       "2088    B7Z8A0  R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...   \n",
       "2089    Q4LE39  R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...   \n",
       "2090    A8K119  R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...   \n",
       "2091    Q6P2M8  R[^DE]{0,2}[^DEPG]([ST])(([FWYLMV].)|([^PRIKGN...   \n",
       "\n",
       "      motif_match          odb_id  odb_mot_st  odb_mot_end  \\\n",
       "0             DPF  10116_0:004cdf         356          358   \n",
       "1             DPF  10116_0:002e5e         399          401   \n",
       "2             DPF  10116_0:002e5e         473          475   \n",
       "3             DPF   9606_0:0016b2         292          294   \n",
       "4             DPF   9606_0:0016b2         297          299   \n",
       "...           ...             ...         ...          ...   \n",
       "2087       RVVSWL   9606_0:003d26          83           88   \n",
       "2088       RINSFE   9606_0:002891         488          493   \n",
       "2089  RSIKSETDSTI   9606_0:0000b6        1058         1068   \n",
       "2090      RSTCNVV   9606_0:002261         246          252   \n",
       "2091      RHSHSGL   9606_0:004e60         411          417   \n",
       "\n",
       "      verified interaction     ELM_motif_class  \n",
       "0                     True      LIG_AP2alpha_2  \n",
       "1                     True      LIG_AP2alpha_2  \n",
       "2                     True      LIG_AP2alpha_2  \n",
       "3                     True      LIG_AP2alpha_2  \n",
       "4                     True      LIG_AP2alpha_2  \n",
       "...                    ...                 ...  \n",
       "2087                 False  LIG_14-3-3_CanoR_1  \n",
       "2088                 False  LIG_14-3-3_CanoR_1  \n",
       "2089                 False  LIG_14-3-3_CanoR_1  \n",
       "2090                 False  LIG_14-3-3_CanoR_1  \n",
       "2091                 False  LIG_14-3-3_CanoR_1  \n",
       "\n",
       "[2092 rows x 11 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elms_df = pd.concat(df_list).reset_index(drop=True)\n",
    "elms_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>ELM_motif_class</th>\n",
       "      <th>DOC_WW_Pin1_4</th>\n",
       "      <th>LIG_14-3-3_CanoR_1</th>\n",
       "      <th>LIG_AP2alpha_2</th>\n",
       "      <th>LIG_EH_1</th>\n",
       "      <th>LIG_SH2_GRB2like</th>\n",
       "      <th>LIG_SH3_CIN85_PxpxPR_1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>verified interaction</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>300</td>\n",
       "      <td>300</td>\n",
       "      <td>299</td>\n",
       "      <td>300</td>\n",
       "      <td>298</td>\n",
       "      <td>298</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>79</td>\n",
       "      <td>52</td>\n",
       "      <td>50</td>\n",
       "      <td>54</td>\n",
       "      <td>30</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "ELM_motif_class       DOC_WW_Pin1_4  LIG_14-3-3_CanoR_1  LIG_AP2alpha_2  \\\n",
       "verified interaction                                                      \n",
       "False                           300                 300             299   \n",
       "True                             79                  52              50   \n",
       "\n",
       "ELM_motif_class       LIG_EH_1  LIG_SH2_GRB2like  LIG_SH3_CIN85_PxpxPR_1  \n",
       "verified interaction                                                      \n",
       "False                      300               298                     298  \n",
       "True                        54                30                      32  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "elms_df.groupby('verified interaction')['ELM_motif_class'].value_counts().unstack()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# manually curated verified interactions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ENAH with [FL]PPPP regex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "from local_orthoDB_group_pipeline import uniprotid_search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get the regex matches for the verified binders first"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "import the annotated binders"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "enah_table = \"../../../../data/manually_curated_interactions/ena_vasp_binders_binding_seq_verification_annotation.xlsx\"\n",
    "df = pd.read_excel(enah_table)\n",
    "df['UniprotID'] = df['UniprotID'].str.strip()\n",
    "verif_ids = df['UniprotID'].unique()\n",
    "verif_id_dict = df[['UniprotID', 'name']].drop_duplicates().set_index('UniprotID').to_dict()['name']\n",
    "assert set(verif_ids) == set(list(verif_id_dict.keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "REGEX = '[FL]PPPP'\n",
    "CLASS_NAME = 'enah_LPPPP_FPPPP'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "import local_env_variables.env_variables as env\n",
    "ORTHODB_DATABASE = env.orthoDB_database()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def uni2odb_ids(uniprot_ids):\n",
    "    odb2uni = {}\n",
    "    for i in uniprot_ids:\n",
    "        try:\n",
    "            odb_id = uniprotid_search.uniprotid_2_odb_gene_id(i)\n",
    "            # uni2odb[i] = odb_id\n",
    "            odb2uni[odb_id] = i\n",
    "        except ValueError as e:\n",
    "            print(e)\n",
    "            print(f\"COULD NOT FIND: {i}, {ligands[i]}\")\n",
    "            continue\n",
    "    return odb2uni\n",
    "\n",
    "\n",
    "def get_verif_table_custom(\n",
    "    uniprot_ids: list,\n",
    "    regex: str,\n",
    "    idr_map: dict|None,\n",
    "    seq_organism: str = \"Homo sapiens\",\n",
    "):\n",
    "    odb2uni = uni2odb_ids(uniprot_ids)\n",
    "    seq_list = [ORTHODB_DATABASE.data_all_seqrecords_dict[i] for i in odb2uni.keys()]\n",
    "    df_verified = regex_search(regex, seq_list, idr_map=idr_map)\n",
    "    df_verified[\"UniprotID\"] = df_verified[\"odb_id\"].map(odb2uni)\n",
    "    df_verified[\"verified interaction\"] = True\n",
    "    df_verified[\"Organism\"] = seq_organism\n",
    "    return df_verified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P51617 not found in gene key table, searching in xref table\n",
      "Q9Y6N7 not found in gene key table, searching in xref table\n",
      "P18206 not found in gene key table, searching in xref table\n",
      "Q70E73 not found in gene key table, searching in xref table\n",
      "Q8WX93 not found in gene key table, searching in xref table\n",
      "Q8TF72 not found in gene key table, searching in xref table\n"
     ]
    }
   ],
   "source": [
    "df_verified = get_verif_table_custom(\n",
    "    verif_ids,\n",
    "    regex=REGEX,\n",
    "    idr_map=idr_map,\n",
    "    seq_organism=\"Homo sapiens\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odb_id</th>\n",
       "      <th>odb_mot_st</th>\n",
       "      <th>odb_mot_end</th>\n",
       "      <th>motif_match</th>\n",
       "      <th>regex</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>verified interaction</th>\n",
       "      <th>Organism</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606_0:0027f1</td>\n",
       "      <td>3253</td>\n",
       "      <td>3257</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q12955</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ANK3_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606_0:0015fb</td>\n",
       "      <td>401</td>\n",
       "      <td>405</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>O15117</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FYB1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606_0:0015fb</td>\n",
       "      <td>625</td>\n",
       "      <td>629</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>O15117</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FYB1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606_0:000e6f</td>\n",
       "      <td>1185</td>\n",
       "      <td>1189</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q9Y6N7</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ROBO1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606_0:000e6f</td>\n",
       "      <td>1479</td>\n",
       "      <td>1483</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q9Y6N7</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ROBO1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>9606_0:002935</td>\n",
       "      <td>841</td>\n",
       "      <td>845</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>P18206</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>VINC_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>9606_0:000f47</td>\n",
       "      <td>23</td>\n",
       "      <td>27</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q702N8</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>XIRP1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>9606_0:001d2c</td>\n",
       "      <td>72</td>\n",
       "      <td>76</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q15942</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZYX_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>9606_0:001d2c</td>\n",
       "      <td>92</td>\n",
       "      <td>96</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q15942</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ZYX_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>9606_0:002a4b</td>\n",
       "      <td>1052</td>\n",
       "      <td>1056</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>O15357</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SHIP2_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>9606_0:00122c</td>\n",
       "      <td>4307</td>\n",
       "      <td>4311</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q14517</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FAT1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>9606_0:00122c</td>\n",
       "      <td>4434</td>\n",
       "      <td>4438</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q14517</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FAT1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>627</td>\n",
       "      <td>631</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>868</td>\n",
       "      <td>872</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>915</td>\n",
       "      <td>919</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>926</td>\n",
       "      <td>930</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>1072</td>\n",
       "      <td>1076</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>9606_0:000b76</td>\n",
       "      <td>1201</td>\n",
       "      <td>1205</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q70E73</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>RAPH1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>9606_0:000661</td>\n",
       "      <td>103</td>\n",
       "      <td>107</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q8WUP2</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FBLI1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>9606_0:000661</td>\n",
       "      <td>165</td>\n",
       "      <td>169</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q8WUP2</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>FBLI1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>9606_0:00141f</td>\n",
       "      <td>561</td>\n",
       "      <td>565</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q8WX93</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PALLD_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>127</td>\n",
       "      <td>131</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>140</td>\n",
       "      <td>144</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>515</td>\n",
       "      <td>519</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>553</td>\n",
       "      <td>557</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>567</td>\n",
       "      <td>571</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>9606_0:00294e</td>\n",
       "      <td>597</td>\n",
       "      <td>601</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q7Z5R6</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>AB1IP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>9606_0:000d90</td>\n",
       "      <td>67</td>\n",
       "      <td>71</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q93052</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>LPP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>9606_0:000d90</td>\n",
       "      <td>87</td>\n",
       "      <td>91</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q93052</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>LPP_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>9606_0:00094d</td>\n",
       "      <td>829</td>\n",
       "      <td>833</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>A6NGG8</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>PCARE_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>9606_0:0012c3</td>\n",
       "      <td>1525</td>\n",
       "      <td>1529</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q8TF72</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SHRM3_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>9606_0:0012c3</td>\n",
       "      <td>1545</td>\n",
       "      <td>1549</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q8TF72</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>SHRM3_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>9606_0:003dae</td>\n",
       "      <td>276</td>\n",
       "      <td>280</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q9P2A4</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ABI3_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>9606_0:003dae</td>\n",
       "      <td>289</td>\n",
       "      <td>293</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q9P2A4</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>ABI3_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>9606_0:001b40</td>\n",
       "      <td>953</td>\n",
       "      <td>957</td>\n",
       "      <td>LPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q5SYE7</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NHSL1_HUMAN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>9606_0:001b40</td>\n",
       "      <td>974</td>\n",
       "      <td>978</td>\n",
       "      <td>FPPPP</td>\n",
       "      <td>[FL]PPPP</td>\n",
       "      <td>Q5SYE7</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NHSL1_HUMAN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           odb_id  odb_mot_st  odb_mot_end motif_match     regex UniprotID  \\\n",
       "0   9606_0:0027f1        3253         3257       FPPPP  [FL]PPPP    Q12955   \n",
       "1   9606_0:0015fb         401          405       LPPPP  [FL]PPPP    O15117   \n",
       "2   9606_0:0015fb         625          629       FPPPP  [FL]PPPP    O15117   \n",
       "3   9606_0:000e6f        1185         1189       LPPPP  [FL]PPPP    Q9Y6N7   \n",
       "4   9606_0:000e6f        1479         1483       LPPPP  [FL]PPPP    Q9Y6N7   \n",
       "5   9606_0:002935         841          845       FPPPP  [FL]PPPP    P18206   \n",
       "6   9606_0:000f47          23           27       LPPPP  [FL]PPPP    Q702N8   \n",
       "7   9606_0:001d2c          72           76       LPPPP  [FL]PPPP    Q15942   \n",
       "8   9606_0:001d2c          92           96       FPPPP  [FL]PPPP    Q15942   \n",
       "9   9606_0:002a4b        1052         1056       FPPPP  [FL]PPPP    O15357   \n",
       "10  9606_0:00122c        4307         4311       LPPPP  [FL]PPPP    Q14517   \n",
       "11  9606_0:00122c        4434         4438       FPPPP  [FL]PPPP    Q14517   \n",
       "12  9606_0:000b76         627          631       LPPPP  [FL]PPPP    Q70E73   \n",
       "13  9606_0:000b76         868          872       FPPPP  [FL]PPPP    Q70E73   \n",
       "14  9606_0:000b76         915          919       FPPPP  [FL]PPPP    Q70E73   \n",
       "15  9606_0:000b76         926          930       FPPPP  [FL]PPPP    Q70E73   \n",
       "16  9606_0:000b76        1072         1076       FPPPP  [FL]PPPP    Q70E73   \n",
       "17  9606_0:000b76        1201         1205       LPPPP  [FL]PPPP    Q70E73   \n",
       "18  9606_0:000661         103          107       LPPPP  [FL]PPPP    Q8WUP2   \n",
       "19  9606_0:000661         165          169       LPPPP  [FL]PPPP    Q8WUP2   \n",
       "20  9606_0:00141f         561          565       FPPPP  [FL]PPPP    Q8WX93   \n",
       "21  9606_0:00294e         127          131       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "22  9606_0:00294e         140          144       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "23  9606_0:00294e         515          519       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "24  9606_0:00294e         553          557       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "25  9606_0:00294e         567          571       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "26  9606_0:00294e         597          601       LPPPP  [FL]PPPP    Q7Z5R6   \n",
       "27  9606_0:000d90          67           71       LPPPP  [FL]PPPP    Q93052   \n",
       "28  9606_0:000d90          87           91       FPPPP  [FL]PPPP    Q93052   \n",
       "29  9606_0:00094d         829          833       LPPPP  [FL]PPPP    A6NGG8   \n",
       "30  9606_0:0012c3        1525         1529       LPPPP  [FL]PPPP    Q8TF72   \n",
       "31  9606_0:0012c3        1545         1549       FPPPP  [FL]PPPP    Q8TF72   \n",
       "32  9606_0:003dae         276          280       LPPPP  [FL]PPPP    Q9P2A4   \n",
       "33  9606_0:003dae         289          293       LPPPP  [FL]PPPP    Q9P2A4   \n",
       "34  9606_0:001b40         953          957       LPPPP  [FL]PPPP    Q5SYE7   \n",
       "35  9606_0:001b40         974          978       FPPPP  [FL]PPPP    Q5SYE7   \n",
       "\n",
       "    verified interaction      Organism         name  \n",
       "0                   True  Homo sapiens   ANK3_HUMAN  \n",
       "1                   True  Homo sapiens   FYB1_HUMAN  \n",
       "2                   True  Homo sapiens   FYB1_HUMAN  \n",
       "3                   True  Homo sapiens  ROBO1_HUMAN  \n",
       "4                   True  Homo sapiens  ROBO1_HUMAN  \n",
       "5                   True  Homo sapiens   VINC_HUMAN  \n",
       "6                   True  Homo sapiens  XIRP1_HUMAN  \n",
       "7                   True  Homo sapiens    ZYX_HUMAN  \n",
       "8                   True  Homo sapiens    ZYX_HUMAN  \n",
       "9                   True  Homo sapiens  SHIP2_HUMAN  \n",
       "10                  True  Homo sapiens   FAT1_HUMAN  \n",
       "11                  True  Homo sapiens   FAT1_HUMAN  \n",
       "12                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "13                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "14                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "15                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "16                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "17                  True  Homo sapiens  RAPH1_HUMAN  \n",
       "18                  True  Homo sapiens  FBLI1_HUMAN  \n",
       "19                  True  Homo sapiens  FBLI1_HUMAN  \n",
       "20                  True  Homo sapiens  PALLD_HUMAN  \n",
       "21                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "22                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "23                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "24                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "25                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "26                  True  Homo sapiens  AB1IP_HUMAN  \n",
       "27                  True  Homo sapiens    LPP_HUMAN  \n",
       "28                  True  Homo sapiens    LPP_HUMAN  \n",
       "29                  True  Homo sapiens  PCARE_HUMAN  \n",
       "30                  True  Homo sapiens  SHRM3_HUMAN  \n",
       "31                  True  Homo sapiens  SHRM3_HUMAN  \n",
       "32                  True  Homo sapiens   ABI3_HUMAN  \n",
       "33                  True  Homo sapiens   ABI3_HUMAN  \n",
       "34                  True  Homo sapiens  NHSL1_HUMAN  \n",
       "35                  True  Homo sapiens  NHSL1_HUMAN  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_verified['name'] = df_verified['UniprotID'].map(verif_id_dict)\n",
    "df_verified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n"
     ]
    }
   ],
   "source": [
    "elm_instances_df = pd.read_csv(ELM_INSTANCES_FILE)\n",
    "df_bg = get_bg_seqs(BG_SEQ_FASTA_FILE, regex=REGEX, instance_df=elm_instances_df, idr_map=idr_map, bg_organism=BG_ORGANISM)\n",
    "df_bg = df_bg.sample(350, random_state=42)\n",
    "dfm = merge_verified_and_fp_motifs(df_verified, df_bg, cluster_dict=BG_CLUSTER_DICT)\n",
    "dfm['ELM_motif_class'] = CLASS_NAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "366\n",
      "2092\n"
     ]
    }
   ],
   "source": [
    "print(len(dfm))\n",
    "print(len(elms_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_df = pd.concat([elms_df, dfm]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TRAF6"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "traf_table = \"../../../../data/manually_curated_interactions/TRAF6_ligands.xlsx\"\n",
    "df = pd.read_excel(traf_table)\n",
    "df['accession'] = df['accession'].str.strip()\n",
    "df = df.rename(columns = {'accession':'UniprotID','motif_seq':'motif_match'})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You could make the regex more specific to reduce the number of false positives<br>\n",
    "It might actually be better to use the information we know about unfavorable residues to make the false positives more likely to be real false positives. That said it's probably safer for now to just use the same regex for both TPs and FPs.\n",
    "- normal regex: `...P.E..[FYWHDE]`\n",
    "- numbering: `(-3)(-2)(-1)(0)(+1)(+2)(+3)(+4)(+5)`\n",
    "- Removing unfavorable residues from the regex:\n",
    "  - proline at any position between (+1) and (+5)\n",
    "    - it binds as a beta sheet\n",
    "  - a positively charged residue (RK) at positions (+3), (+4), or (+5)\n",
    "\n",
    "\n",
    "**To reduce the number of false positives, I am just going to subsample the false positives**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "# REGEX='...P[^P]E[^PRK][^PRK][FYWDE]'\n",
    "REGEX='...P.E..[FYWDE]'\n",
    "CLASS_NAME='TRAF6'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "def odb_id_2_species_name(odb_gene_id: str):\n",
    "    spec_id=sql_queries.odb_gene_id_2_species_id(odb_gene_id)\n",
    "    return ORTHODB_DATABASE.data_species_dict[spec_id]\n",
    "\n",
    "\n",
    "def get_verif_table_custom2(\n",
    "    uniprot_ids: list,\n",
    "    regex: str,\n",
    "):\n",
    "    odb2uni = uni2odb_ids(uniprot_ids)\n",
    "    seq_list = [ORTHODB_DATABASE.data_all_seqrecords_dict[i] for i in odb2uni.keys()]\n",
    "    df_verified = regex_search(regex, seq_list, idr_map=None)\n",
    "    df_verified[\"UniprotID\"] = df_verified[\"odb_id\"].map(odb2uni)\n",
    "    df_verified[\"verified interaction\"] = True\n",
    "    df_verified[\"Organism\"] = df_verified[\"odb_id\"].apply(odb_id_2_species_name)\n",
    "    return df_verified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "P25942 not found in gene key table, searching in xref table\n",
      "P51617 not found in gene key table, searching in xref table\n"
     ]
    }
   ],
   "source": [
    "df_verified = get_verif_table_custom2(\n",
    "    list(df['UniprotID'].unique()),\n",
    "    regex=REGEX,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>motif_match</th>\n",
       "      <th>name</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P25942</td>\n",
       "      <td>KQEPQEINF</td>\n",
       "      <td>CD40</td>\n",
       "      <td>1999-Tsukamoto, 1999-Pullen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q96CG3</td>\n",
       "      <td>SSSPTEMDE</td>\n",
       "      <td>TIFA</td>\n",
       "      <td>2019-Huang</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7Z434</td>\n",
       "      <td>CHGPEENEY</td>\n",
       "      <td>MAVS</td>\n",
       "      <td>2015-Shi</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8IUC6</td>\n",
       "      <td>CQEPEEMSW</td>\n",
       "      <td>TICAM1</td>\n",
       "      <td>2004-Jiang, 2003-Sato</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O43187</td>\n",
       "      <td>SNTPEETDD</td>\n",
       "      <td>IRAK2</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P51617</td>\n",
       "      <td>PPSPQENSY</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P51617</td>\n",
       "      <td>PNQPVESDE</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P51617</td>\n",
       "      <td>RQGPEESDE</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q9Y616</td>\n",
       "      <td>PSIPVEDDE</td>\n",
       "      <td>IRAK3 (IRAK-M)</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>O35305</td>\n",
       "      <td>RKIPTEDEY</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>O35305</td>\n",
       "      <td>FQEPLEVGE</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>O35305</td>\n",
       "      <td>GNTPGEDHE</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   UniprotID motif_match                    name                       Source\n",
       "0     P25942   KQEPQEINF                    CD40  1999-Tsukamoto, 1999-Pullen\n",
       "1     Q96CG3   SSSPTEMDE                    TIFA                   2019-Huang\n",
       "2     Q7Z434   CHGPEENEY                    MAVS                     2015-Shi\n",
       "3     Q8IUC6   CQEPEEMSW                  TICAM1        2004-Jiang, 2003-Sato\n",
       "4     O43187   SNTPEETDD                   IRAK2                      2002-Ye\n",
       "5     P51617   PPSPQENSY                   IRAK1                      2002-Ye\n",
       "6     P51617   PNQPVESDE                   IRAK1                      2002-Ye\n",
       "7     P51617   RQGPEESDE                   IRAK1                      2002-Ye\n",
       "8     Q9Y616   PSIPVEDDE          IRAK3 (IRAK-M)                      2002-Ye\n",
       "9     O35305   RKIPTEDEY  mouse TNFRSF11A (RANK)                      2002-Ye\n",
       "10    O35305   FQEPLEVGE  mouse TNFRSF11A (RANK)                      2002-Ye\n",
       "11    O35305   GNTPGEDHE  mouse TNFRSF11A (RANK)                      2002-Ye"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "merge the regex matches with the annotated binders to restrict the \"verified\" interactions to only those specific sequences in the original excel file. For this protein, I don't just want to take the regex matches from proteins that are known to bind, I want the specific sequences that are known to bind.<br>\n",
    "\n",
    "You could say that I should have done the same thing for enah, but I think the fact that I used the [LF]PPPP regex will make the matches very very likely to be real."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_verified=pd.merge(left=df, right=df_verified, how='left', on=['UniprotID', 'motif_match'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>motif_match</th>\n",
       "      <th>name</th>\n",
       "      <th>Source</th>\n",
       "      <th>odb_id</th>\n",
       "      <th>odb_mot_st</th>\n",
       "      <th>odb_mot_end</th>\n",
       "      <th>regex</th>\n",
       "      <th>verified interaction</th>\n",
       "      <th>Organism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>P25942</td>\n",
       "      <td>KQEPQEINF</td>\n",
       "      <td>CD40</td>\n",
       "      <td>1999-Tsukamoto, 1999-Pullen</td>\n",
       "      <td>9606_0:004882</td>\n",
       "      <td>233</td>\n",
       "      <td>241</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Q96CG3</td>\n",
       "      <td>SSSPTEMDE</td>\n",
       "      <td>TIFA</td>\n",
       "      <td>2019-Huang</td>\n",
       "      <td>9606_0:001440</td>\n",
       "      <td>172</td>\n",
       "      <td>180</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Q7Z434</td>\n",
       "      <td>CHGPEENEY</td>\n",
       "      <td>MAVS</td>\n",
       "      <td>2015-Shi</td>\n",
       "      <td>9606_0:00486f</td>\n",
       "      <td>451</td>\n",
       "      <td>459</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Q8IUC6</td>\n",
       "      <td>CQEPEEMSW</td>\n",
       "      <td>TICAM1</td>\n",
       "      <td>2004-Jiang, 2003-Sato</td>\n",
       "      <td>9606_0:004368</td>\n",
       "      <td>246</td>\n",
       "      <td>254</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>O43187</td>\n",
       "      <td>SNTPEETDD</td>\n",
       "      <td>IRAK2</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>9606_0:000e31</td>\n",
       "      <td>522</td>\n",
       "      <td>530</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>P51617</td>\n",
       "      <td>PPSPQENSY</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>9606_0:004fa3</td>\n",
       "      <td>538</td>\n",
       "      <td>546</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>P51617</td>\n",
       "      <td>PNQPVESDE</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>9606_0:004fa3</td>\n",
       "      <td>581</td>\n",
       "      <td>589</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>P51617</td>\n",
       "      <td>RQGPEESDE</td>\n",
       "      <td>IRAK1</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>9606_0:004fa3</td>\n",
       "      <td>700</td>\n",
       "      <td>708</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Q9Y616</td>\n",
       "      <td>PSIPVEDDE</td>\n",
       "      <td>IRAK3 (IRAK-M)</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>9606_0:0031e9</td>\n",
       "      <td>474</td>\n",
       "      <td>482</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Homo sapiens</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>O35305</td>\n",
       "      <td>RKIPTEDEY</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>10090_0:000361</td>\n",
       "      <td>336</td>\n",
       "      <td>344</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>O35305</td>\n",
       "      <td>FQEPLEVGE</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>10090_0:000361</td>\n",
       "      <td>369</td>\n",
       "      <td>377</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>O35305</td>\n",
       "      <td>GNTPGEDHE</td>\n",
       "      <td>mouse TNFRSF11A (RANK)</td>\n",
       "      <td>2002-Ye</td>\n",
       "      <td>10090_0:000361</td>\n",
       "      <td>443</td>\n",
       "      <td>451</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>True</td>\n",
       "      <td>Mus musculus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   UniprotID motif_match                    name                       Source  \\\n",
       "0     P25942   KQEPQEINF                    CD40  1999-Tsukamoto, 1999-Pullen   \n",
       "1     Q96CG3   SSSPTEMDE                    TIFA                   2019-Huang   \n",
       "2     Q7Z434   CHGPEENEY                    MAVS                     2015-Shi   \n",
       "3     Q8IUC6   CQEPEEMSW                  TICAM1        2004-Jiang, 2003-Sato   \n",
       "4     O43187   SNTPEETDD                   IRAK2                      2002-Ye   \n",
       "5     P51617   PPSPQENSY                   IRAK1                      2002-Ye   \n",
       "6     P51617   PNQPVESDE                   IRAK1                      2002-Ye   \n",
       "7     P51617   RQGPEESDE                   IRAK1                      2002-Ye   \n",
       "8     Q9Y616   PSIPVEDDE          IRAK3 (IRAK-M)                      2002-Ye   \n",
       "9     O35305   RKIPTEDEY  mouse TNFRSF11A (RANK)                      2002-Ye   \n",
       "10    O35305   FQEPLEVGE  mouse TNFRSF11A (RANK)                      2002-Ye   \n",
       "11    O35305   GNTPGEDHE  mouse TNFRSF11A (RANK)                      2002-Ye   \n",
       "\n",
       "            odb_id  odb_mot_st  odb_mot_end            regex  \\\n",
       "0    9606_0:004882         233          241  ...P.E..[FYWDE]   \n",
       "1    9606_0:001440         172          180  ...P.E..[FYWDE]   \n",
       "2    9606_0:00486f         451          459  ...P.E..[FYWDE]   \n",
       "3    9606_0:004368         246          254  ...P.E..[FYWDE]   \n",
       "4    9606_0:000e31         522          530  ...P.E..[FYWDE]   \n",
       "5    9606_0:004fa3         538          546  ...P.E..[FYWDE]   \n",
       "6    9606_0:004fa3         581          589  ...P.E..[FYWDE]   \n",
       "7    9606_0:004fa3         700          708  ...P.E..[FYWDE]   \n",
       "8    9606_0:0031e9         474          482  ...P.E..[FYWDE]   \n",
       "9   10090_0:000361         336          344  ...P.E..[FYWDE]   \n",
       "10  10090_0:000361         369          377  ...P.E..[FYWDE]   \n",
       "11  10090_0:000361         443          451  ...P.E..[FYWDE]   \n",
       "\n",
       "    verified interaction      Organism  \n",
       "0                   True  Homo sapiens  \n",
       "1                   True  Homo sapiens  \n",
       "2                   True  Homo sapiens  \n",
       "3                   True  Homo sapiens  \n",
       "4                   True  Homo sapiens  \n",
       "5                   True  Homo sapiens  \n",
       "6                   True  Homo sapiens  \n",
       "7                   True  Homo sapiens  \n",
       "8                   True  Homo sapiens  \n",
       "9                   True  Mus musculus  \n",
       "10                  True  Mus musculus  \n",
       "11                  True  Mus musculus  "
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_verified"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n",
      "added cluster members\n"
     ]
    }
   ],
   "source": [
    "elm_instances_df = pd.read_csv(ELM_INSTANCES_FILE)\n",
    "df_bg = get_bg_seqs(\n",
    "    BG_SEQ_FASTA_FILE,\n",
    "    regex=REGEX,\n",
    "    instance_df=elm_instances_df,\n",
    "    idr_map=idr_map,\n",
    "    bg_organism=BG_ORGANISM,\n",
    ")\n",
    "df_bg = df_bg.sample(300, random_state=42)\n",
    "dfm = merge_verified_and_fp_motifs(df_verified, df_bg, cluster_dict=BG_CLUSTER_DICT)\n",
    "dfm[\"ELM_motif_class\"] = CLASS_NAME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "312\n"
     ]
    }
   ],
   "source": [
    "print(len(dfm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_df = pd.concat([benchmark_df, dfm]).reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organism</th>\n",
       "      <th>Primary_Acc</th>\n",
       "      <th>Accessions</th>\n",
       "      <th>UniprotID</th>\n",
       "      <th>regex</th>\n",
       "      <th>motif_match</th>\n",
       "      <th>odb_id</th>\n",
       "      <th>odb_mot_st</th>\n",
       "      <th>odb_mot_end</th>\n",
       "      <th>verified interaction</th>\n",
       "      <th>ELM_motif_class</th>\n",
       "      <th>name</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>O08838</td>\n",
       "      <td>O08838</td>\n",
       "      <td>O08838</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:004cdf</td>\n",
       "      <td>356</td>\n",
       "      <td>358</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>399</td>\n",
       "      <td>401</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>Q05140</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>10116_0:002e5e</td>\n",
       "      <td>473</td>\n",
       "      <td>475</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>P98082 A6NES5 Q13598 Q9BTY0 Q9UK04</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>292</td>\n",
       "      <td>294</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>P98082</td>\n",
       "      <td>P98082 A6NES5 Q13598 Q9BTY0 Q9UK04</td>\n",
       "      <td>P98082</td>\n",
       "      <td>DP[FW]</td>\n",
       "      <td>DPF</td>\n",
       "      <td>9606_0:0016b2</td>\n",
       "      <td>297</td>\n",
       "      <td>299</td>\n",
       "      <td>True</td>\n",
       "      <td>LIG_AP2alpha_2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2765</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B4DHE8</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>KAQPKEVMF</td>\n",
       "      <td>9606_0:00409b</td>\n",
       "      <td>183</td>\n",
       "      <td>191</td>\n",
       "      <td>False</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2766</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>B3KU03</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>EPIPQEPQW</td>\n",
       "      <td>9606_0:000730</td>\n",
       "      <td>764</td>\n",
       "      <td>772</td>\n",
       "      <td>False</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2767</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E9PEI6</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>TPSPAEPTE</td>\n",
       "      <td>9606_0:0019f6</td>\n",
       "      <td>392</td>\n",
       "      <td>400</td>\n",
       "      <td>False</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2768</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>O94888</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>LEMPPEKAD</td>\n",
       "      <td>9606_0:000f70</td>\n",
       "      <td>391</td>\n",
       "      <td>399</td>\n",
       "      <td>False</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2769</th>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q658X5</td>\n",
       "      <td>...P.E..[FYWDE]</td>\n",
       "      <td>NPKPEELRE</td>\n",
       "      <td>9606_0:003617</td>\n",
       "      <td>614</td>\n",
       "      <td>622</td>\n",
       "      <td>False</td>\n",
       "      <td>TRAF6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2770 rows Ã— 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               Organism Primary_Acc                          Accessions  \\\n",
       "0     Rattus norvegicus      O08838                              O08838   \n",
       "1     Rattus norvegicus      Q05140                              Q05140   \n",
       "2     Rattus norvegicus      Q05140                              Q05140   \n",
       "3          Homo sapiens      P98082  P98082 A6NES5 Q13598 Q9BTY0 Q9UK04   \n",
       "4          Homo sapiens      P98082  P98082 A6NES5 Q13598 Q9BTY0 Q9UK04   \n",
       "...                 ...         ...                                 ...   \n",
       "2765       Homo sapiens         NaN                                 NaN   \n",
       "2766       Homo sapiens         NaN                                 NaN   \n",
       "2767       Homo sapiens         NaN                                 NaN   \n",
       "2768       Homo sapiens         NaN                                 NaN   \n",
       "2769       Homo sapiens         NaN                                 NaN   \n",
       "\n",
       "     UniprotID            regex motif_match          odb_id  odb_mot_st  \\\n",
       "0       O08838           DP[FW]         DPF  10116_0:004cdf         356   \n",
       "1       Q05140           DP[FW]         DPF  10116_0:002e5e         399   \n",
       "2       Q05140           DP[FW]         DPF  10116_0:002e5e         473   \n",
       "3       P98082           DP[FW]         DPF   9606_0:0016b2         292   \n",
       "4       P98082           DP[FW]         DPF   9606_0:0016b2         297   \n",
       "...        ...              ...         ...             ...         ...   \n",
       "2765    B4DHE8  ...P.E..[FYWDE]   KAQPKEVMF   9606_0:00409b         183   \n",
       "2766    B3KU03  ...P.E..[FYWDE]   EPIPQEPQW   9606_0:000730         764   \n",
       "2767    E9PEI6  ...P.E..[FYWDE]   TPSPAEPTE   9606_0:0019f6         392   \n",
       "2768    O94888  ...P.E..[FYWDE]   LEMPPEKAD   9606_0:000f70         391   \n",
       "2769    Q658X5  ...P.E..[FYWDE]   NPKPEELRE   9606_0:003617         614   \n",
       "\n",
       "      odb_mot_end  verified interaction ELM_motif_class name Source  \n",
       "0             358                  True  LIG_AP2alpha_2  NaN    NaN  \n",
       "1             401                  True  LIG_AP2alpha_2  NaN    NaN  \n",
       "2             475                  True  LIG_AP2alpha_2  NaN    NaN  \n",
       "3             294                  True  LIG_AP2alpha_2  NaN    NaN  \n",
       "4             299                  True  LIG_AP2alpha_2  NaN    NaN  \n",
       "...           ...                   ...             ...  ...    ...  \n",
       "2765          191                 False           TRAF6  NaN    NaN  \n",
       "2766          772                 False           TRAF6  NaN    NaN  \n",
       "2767          400                 False           TRAF6  NaN    NaN  \n",
       "2768          399                 False           TRAF6  NaN    NaN  \n",
       "2769          622                 False           TRAF6  NaN    NaN  \n",
       "\n",
       "[2770 rows x 13 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmark_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# filter out really long sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_SEQ_LENGTH = 5000\n",
    "# change to 4000 in the next iteration\n",
    "# MAX_SEQ_LENGTH = 4000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_seq_length(odb_gene_id: str) -> int:\n",
    "    return len(ORTHODB_DATABASE.data_all_seqrecords_dict[odb_gene_id].seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_df['seq_length'] = benchmark_df['odb_id'].apply(get_seq_length)\n",
    "benchmark_df = benchmark_df[benchmark_df['seq_length'] < MAX_SEQ_LENGTH]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2715\n"
     ]
    }
   ],
   "source": [
    "print(len(benchmark_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>verified interaction</th>\n",
       "      <th>False</th>\n",
       "      <th>True</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ELM_motif_class</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>DOC_WW_Pin1_4</th>\n",
       "      <td>288</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIG_14-3-3_CanoR_1</th>\n",
       "      <td>287</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIG_AP2alpha_2</th>\n",
       "      <td>289</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIG_EH_1</th>\n",
       "      <td>297</td>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIG_SH2_GRB2like</th>\n",
       "      <td>291</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIG_SH3_CIN85_PxpxPR_1</th>\n",
       "      <td>297</td>\n",
       "      <td>32</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TRAF6</th>\n",
       "      <td>292</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>enah_LPPPP_FPPPP</th>\n",
       "      <td>329</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "verified interaction    False  True \n",
       "ELM_motif_class                     \n",
       "DOC_WW_Pin1_4             288     79\n",
       "LIG_14-3-3_CanoR_1        287     52\n",
       "LIG_AP2alpha_2            289     50\n",
       "LIG_EH_1                  297     54\n",
       "LIG_SH2_GRB2like          291     30\n",
       "LIG_SH3_CIN85_PxpxPR_1    297     32\n",
       "TRAF6                     292     12\n",
       "enah_LPPPP_FPPPP          329     36"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "benchmark_df.groupby('ELM_motif_class')['verified interaction'].value_counts().unstack().fillna(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2080"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(benchmark_df['odb_id'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "rn_dict = {\n",
    "    'motif_match': 'hit_sequence',\n",
    "    'odb_id': 'gene_id',\n",
    "    'odb_mot_st': 'hit start position',\n",
    "    'odb_mot_end': 'hit end position',\n",
    "}\n",
    "\n",
    "benchmark_df = benchmark_df.rename(columns=rn_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "benchmark_df.to_csv(BENCHMARK_DIR / \"benchmark_table.csv\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# TODO\n",
    "- [ ] add LIR to benchmark\n",
    "- [ ] Change the column names **here** to prepare for the conservation pipeline and avoid running that renaming script"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "slim_conservation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
